internal/kallsyms/kallsyms.go

```diff
@@ -72,3 +78,55 @@ func loadKernelModuleMapping(f io.Reader) (map[string]string, error) {
 	}
 	return mods, nil
 }
+
+func LoadSymbolAddresses() (map[string]uint64, error) {
+	kernelSymbols.RLock()
+	ksyms := kernelSymbols.ksyms
+	kernelSymbols.RUnlock()
+
+	if ksyms == nil {
+		kernelSymbols.Lock()
+		defer kernelSymbols.Unlock()
+		ksyms = kernelSymbols.ksyms
+	}
+
+	f, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return nil, err
+	}
+
+	ksyms, err = loadSymbolAddresses(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return ksyms, nil
+}
+
+func loadSymbolAddresses(f io.Reader) (map[string]uint64, error) {
```

Comment:
I am not yet 100% convinced this is the way to go. This essentially loads all symbols into memory, persistently. A quick dump on my local machine indicates `/proc/kallsyms` is about 17MiB (it will vary depending on the amount of modules loaded),  we are removing a bit due to having the address as uint64 and adding a bit due to map overhead. So it likely isn't super far off.

My thinking is, `/proc/kallsyms` is a virtual file, so reading it is likely faster than reading an actual file. Memory allocation is slow, especially maps and especially when you can't provide a capacity. Might it not be better to search the `/proc/kallsyms` instead of pre-indexing everything. Both time wise and memory wise.

I would at least like to do a bit of research before committing to one or the other approach.

Reply:
That is a good point. I had another look how libbpf is handling this and they are patching the address in the symbols directly. I guess a better approach would be to simply collect all used symbols into a map, iterate over the data of `/proc/kallsym` and then patch the addresses from there. What do you think?

Reply:
I did some quick testing.

Added benchmarks for the map indexing and searching approach:
```
func BenchmarkLoadSymbolAddresses(b *testing.B) {
	for i := 0; i < b.N; i++ {
		f, err := os.Open("/proc/kallsyms")
		if err != nil {
			b.Fatal(err)
		}

		_, err = loadSymbolAddresses(f)
		if err != nil {
			b.Fatal(err)
		}
	}
}

func BenchmarkFindSymbolAddresses(b *testing.B) {
	for i := 0; i < b.N; i++ {
		f, err := os.Open("/proc/kallsyms")
		if err != nil {
			b.Fatal(err)
		}

		_, err = findSymbolAddresses(f, []string{
			"fixed_percpu_data",
			"__per_cpu_start",
			"cpu_debug_store",
			"irq_stack_backing_store",
			"cpu_tss_rw",
			"gdt_page",
			"exception_stacks",
			"entry_stack_storage",
			"espfix_waddr",
			"espfix_stack",
		})
		if err != nil {
			b.Fatal(err)
		}
	}
}

func TestFindSymbolAddresses(t *testing.T) {
	f, err := os.Open("/proc/kallsyms")
	if err != nil {
		t.Fatal(err)
	}

	_, err = findSymbolAddresses(f, []string{
		"fixed_percpu_data",
		"__per_cpu_start",
		"cpu_debug_store",
		"irq_stack_backing_store",
		"cpu_tss_rw",
		"gdt_page",
		"exception_stacks",
		"entry_stack_storage",
		"espfix_waddr",
		"espfix_stack",
	})
	if err != nil {
		t.Fatal(err)
	}

	t.Fail()
}
```

Implemented a rough version which minimizes allocations:
```
func findSymbolAddresses(f io.Reader, syms []string) ([]uint64, error) {
	r := bufio.NewReader(f)
	result := make([]uint64, len(syms))

	const bufSize = 1024 * 16
	buf := make([]byte, bufSize)
	var err error
	i := 0
	n := 0
	for {
		newLine := bytes.IndexByte(buf[i:n], '\n')
		if newLine == -1 {
			rem := buf[i:n]
			copy(buf, rem)

			n, err = r.Read(buf[len(rem):])
			if err != nil {
				if err == io.EOF {
					return result, nil
				}
				return nil, err
			}
			n = n + len(rem)
			i = 0

			continue
		}

		line := buf[i : i+newLine]

		sIdx := bytes.IndexByte(line, ' ')
		if sIdx == -1 {
			continue
		}
		addrBytes := line[:sIdx]
		line = line[sIdx+1:]

		sIdx = bytes.IndexByte(line, ' ')
		if sIdx == -1 {
			continue
		}
		sIdx = bytes.IndexByte(line, ' ')
		symBytes := line
		if sIdx != -1 {
			symBytes = line[:sIdx]
		}

		for i, sym := range syms {
			if string(symBytes) == sym {
				addr, err := strconv.ParseUint(string(addrBytes), 16, 64)
				if err != nil {
					return nil, err
				}
				result[i] = addr
			}
		}

		i += newLine + 1
	}
}
```

Results on my machine are:
```
go test -benchmem -run=^$ -bench=".*" ./internal/kallsyms 
goos: linux
goarch: amd64
pkg: github.com/cilium/ebpf/internal/kallsyms
cpu: 13th Gen Intel(R) Core(TM) i7-13800H
BenchmarkLoadSymbolAddresses-20                2         631965570 ns/op        101882808 B/op   2936085 allocs/op
BenchmarkFindSymbolAddresses-20                8         126794984 ns/op           20682 B/op          6 allocs/op
PASS
ok      github.com/cilium/ebpf/internal/kallsyms        3.067s
```

The benchmark searches 10 symbols, but varying this does not seem to change much performance wise. A pprof shows 99% in the read syscall. So this is near the best we can do by re-parsing every time.

So there is an interesting tradeoff. Pre-indexing starts to become faster after 5 calls (regardless of the amount of symbols). At 4 calls CPU time is even, but then re-parsing still wins on memory usage. Makes me wonder if it would be possible to collect the symbols for all programs in a collection and fixup all programs at once?

Reply:
Thanks for sharing those result! 

I've updated the code and now parse `/proc/kallsyms` once during loading of the collection. I didn't update the parser function for the individual lines yet, but would love to hear what you think about the updated solution.

Reply:
Thank you! This approach makes more sense to me. We are still a bit slow in `loadSymbolAddresses`, that is simply because the bufio scanner, and scanf do a lot of allocation internally. Lets deal with that some other time (another PR). 

However, even with improved performance walking the kallsyms is slow read wise. Can we make calling `LoadSymbolAddresses` conditional, so we only take the hit if we have at least 1 ksym to resolve?

Reply:
I will add the check 👌 We can also return as soon as we have found all symbols, no need to iterate over potentially the full file.

Reply:
> We can also return as soon as we have found all symbols, no need to iterate over potentially the full file.

Yea, also a very good idea. :+1:

Reply:
I've implemented the optimization and cleanup the commit history. I will now also turn this back to a PR from the draft 👍 

Thanks for the nice pointers 👌